File: lectures/lecture09/python/list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None  # only used for DoublyLinkedList


class SinglyLinkedList:
    """
    singly linked list (generic)
    """

    def __init__(self):
        self.head = None

    def push_front(self, data):
        """
        add element to front of list, O(1)
        """
        if not self.head:
            self.head = Node(data)
        else:
            node = Node(data)
            node.next = self.head
            self.head = node

    def pop_front(self):
        """
        pop element from front of list, O(1)
        """
        if not self.head:
            raise Exception("Cannot pop from empty list")
        data = self.head.data
        self.head = self.head.next
        return data

    def push_back(self, data):
        """
        add element to back of list, O(N)
        """
        if not self.head:
            self.head = Node(data)
        else:
            node = self.head
            while node.next is not None:
                node = node.next
            node.next = Node(data)

    def pop_back(self):
        """
        pop element from back of list, O(N)
        """
        if not self.head:
            raise Exception("Cannot pop from empty list")
        elif not self.head.next:
            data = self.head.data
            self.head = None
            return data
        else:
            node = self.head
            while node.next.next is not None:
                node = node.next
            data = node.next.data
            node.next = None
            return data

    def __str__(self):
        """
        string representation of list
        """
        s = ""
        node = self.head
        while node is not None:
            s += str(node.data) + " "
            node = node.next
        return s


class DoublyLinkedList:
    """
    doubly linked list (generic)
    """

    def __init__(self):
        self.head = None
        self.tail = None

    def push_front(self, data):
        """
        add element to front of list, O(1)
        """
        if not self.head:
            self.head = Node(data)
            self.tail = self.head
        else:
            node = Node(data)
            node.next = self.head
            self.head.prev = node
            self.head = node

    def pop_front(self):
        """
        pop element from front of list, O(1)
        """
        if not self.head:
            raise Exception("Cannot pop from empty list")
        data = self.head.data
        self.head = self.head.next
        if not self.head:
            self.tail = None
        else:
            self.head.prev = None
        return data

    def push_back(self, data):
        """
        add element to back of list, O(1)
        """
        if not self.tail:
            self.tail = Node(data)
            self.head = self.tail
        else:
            node = Node(data)
            self.tail.next = node
            node.prev = self.tail
            self.tail = node

    def pop_back(self):
        """
        pop element from back of list, O(1)
        """
        if not self.tail:
            raise Exception("Cannot pop from empty list")
        data = self.tail.data
        self.tail = self.tail.prev
        if not self.tail:
            self.head = None
        else:
            self.tail.next = None
        return data

    def __str__(self):
        """
        string representation of list
        """
        s = ""
        node = self.head
        while node is not None:
            s += str(node.data) + " "
            node = node.next
        return s


if __name__ == "__main__":

    linked_list = SinglyLinkedList()
    linked_list.push_front(5)
    linked_list.push_front(3)
    linked_list.push_front(7)
    print(str(linked_list))  # 7, 3, 5
    linked_list.pop_front()
    print(str(linked_list))  # 3, 5

    linked_list = SinglyLinkedList()
    linked_list.push_back(5)
    linked_list.push_back(3)
    linked_list.push_back(7)
    print(str(linked_list))  # 5, 3, 7
    linked_list.pop_back()
    print(str(linked_list))  # 5, 3

    linked_list = DoublyLinkedList()
    linked_list.push_front(5)
    linked_list.push_front(3)
    linked_list.push_front(7)
    print(str(linked_list))  # 7, 3, 5
    linked_list.pop_front()
    print(str(linked_list))  # 3, 5
    linked_list.pop_front()
    linked_list.pop_front()

    linked_list = DoublyLinkedList()
    linked_list.push_back(5)
    linked_list.push_back(3)
    linked_list.push_back(7)
    print(str(linked_list))  # 5, 3, 7
    linked_list.pop_back()
    print(str(linked_list))  # 5, 3
    linked_list.pop_back()
    linked_list.pop_back()

File: courses/cmpt225/lecture09/cpp/src/list.cpp
#include "list.h"
#include "node.h"

#include <exception>
#include <string>

template <class Type>
SinglyLinkedList<Type>::SinglyLinkedList() : head(nullptr)
{
}

template <class Type>
SinglyLinkedList<Type>::~SinglyLinkedList()
{
    Node<Type> *node = head;
    while (node) {
        Node<Type> *next = node->next;
        delete node;
        node = next;
    }
}

template <class Type>
void SinglyLinkedList<Type>::push_front(Type data)
{
    if (!head)
    {
        head = new Node<Type>(data);
    }
    else
    {
        Node<Type> *node = new Node<Type>(data);
        node->next = head;
        head = node;
    }
}

template <class Type>
Type SinglyLinkedList<Type>::pop_front()
{
    if (!head)
    {
        throw std::exception("Cannot pop from empty list");
    }

    Type data = head->data;
    Node<Type> *next = head->next;
    delete head;
    head = next;

    return data;
}

template <class Type>
void SinglyLinkedList<Type>::push_back(Type data)
{
    if (!head)
    {
        head = new Node<Type>(data);
    }
    else
    {
        Node<Type> *node = head;
        while (node->next) {
            node = node->next;
        }
        node->next = new Node<Type>(data);
    }
}

template <class Type>
Type SinglyLinkedList<Type>::pop_back()
{
    if (!head)
    {
        throw std::exception("Cannot pop from empty list");
    }
    else if (!head->next)
    {
        Type data = head->data;
        delete head;
        head = nullptr;
        return data;
    }
    else
    {
        Node<Type> *node = head;
        while (node->next->next)
        {
            node = node->next;
        }
        Type data = node->next->data;
        delete node->next;
        node->next = nullptr;
        return data;
    }
}

std::string SinglyLinkedList<int>::to_string() const
{
    std::string s;

    Node<int> *node = head;

    while (node)
    {
        s += std::to_string(node->data) + " ";
        node = node->next;
    }

    return s;
}

std::string SinglyLinkedList<std::string>::to_string() const
{
    std::string s;

    Node<std::string> *node = head;

    while (node)
    {
        s += node->data + " ";
        node = node->next;
    }

    return s;
}

template <class Type>
DoublyLinkedList<Type>::DoublyLinkedList() : head(nullptr)
{
}

template <class Type>
DoublyLinkedList<Type>::~DoublyLinkedList()
{
    Node<Type> *node = head;
    while (node) {
        Node<Type> *next = node->next;
        delete node;
        node = next;
    }
}

template <class Type>
void DoublyLinkedList<Type>::push_front(Type data)
{
    if (!head)
    {
        head = new Node<Type>(data);
        tail = head;
    }
    else
    {
        Node<Type> *node = new Node<Type>(data);
        node->next = head;
        head->prev = node;
        head = node;
    }
}

template <class Type>
Type DoublyLinkedList<Type>::pop_front()
{
    if (!head)
    {
        throw std::exception("Cannot pop from empty list");
    }

    Type data = head->data;
    Node<Type> *next = head->next;
    delete head;
    head = next;

    if (!head)
    {
        tail = nullptr;
    }
    else
    {
        head->prev = nullptr;
    }

    return data;
}

template <class Type>
void DoublyLinkedList<Type>::push_back(Type data)
{
    if (!tail)
    {
        tail = new Node<Type>(data);
        head = tail;
    }
    else
    {
        Node<Type> *node = new Node<Type>(data);
        tail->next = node;
        node->prev = tail;
        tail = node;
    }
}

template <class Type>
Type DoublyLinkedList<Type>::pop_back()
{
    if (!tail)
    {
        throw std::exception("Cannot pop from empty list");
    }

    Type data = tail->data;
    Node<Type> *prev = tail->prev;
    delete tail;
    tail = tail->prev;

    if (!tail)
    {
        head = nullptr;
    }
    else
    {
        tail->next = nullptr;
    }

    return data; 
}

std::string DoublyLinkedList<int>::to_string() const
{
    std::string s;

    Node<int> *node = head;

    while (node)
    {
        s += std::to_string(node->data) + " ";
        node = node->next;
    }

    return s;
}

std::string DoublyLinkedList<std::string>::to_string() const
{
    std::string s;

    Node<std::string> *node = head;

    while (node)
    {
        s += node->data + " ";
        node = node->next;
    }

    return s;
}

template class SinglyLinkedList<int>;
template class SinglyLinkedList<std::string>;

template class DoublyLinkedList<int>;
template class DoublyLinkedList<std::string>;
